runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,339 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
 TR::Register *
 J9::Z::TreeEvaluator::multianewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    {
-   TR::ILOpCodes opCode = node->getOpCodeValue();
-   TR::Node::recreate(node, TR::acall);
-   TR::Register * targetRegister = directCallEvaluator(node, cg);
-   TR::Node::recreate(node, opCode);
-   return targetRegister;
-   }
+   static char *useDirectHelperCall = feGetEnv("TR_MultiANewArrayEvaluatorUseDirectCall");
+   TR::Compilation *comp = cg->comp();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(comp->fe());
+   TR::Register *targetReg = cg->allocateRegister();
+
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Node *thirdChild = node->getThirdChild();
+
+   if (useDirectHelperCall || !secondChild->getOpCode().isLoadConst() || secondChild->getInt()!=2)
+      {
+      traceMsg(cg->comp(), "DCDCDCDC - multianewArrayEvaluator. More than 2 dimensions, so we can't handle this scenario in JITT'ed code\n");
+      return performHelperCall(node, NULL, TR::acall, false, cg);
+      }
+
+   TR::LabelSymbol *cFlowRegionStart = generateLabelSymbol(cg);
+   TR::LabelSymbol *nonZeroFirstDimLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *cFlowRegionEnd = generateLabelSymbol(cg);
+   TR::LabelSymbol *oolFailLabel = generateLabelSymbol(cg);
+
+   cFlowRegionStart->setStartInternalControlFlow();
+   cFlowRegionEnd->setEndInternalControlFlow();
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionStart);
+
+   TR::Register *dimsPtrReg = cg->evaluate(firstChild);
+   TR::Register *dimReg = cg->evaluate(secondChild);
+   TR::Register *classReg = cg->evaluate(thirdChild);
+
+   // First generate the OOL code. This path will be used to handle the scenarios that we don't want to handle in JITT'ed code.
+   TR_S390OutOfLineCodeSection *outlinedSlowPath = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(oolFailLabel, cFlowRegionEnd, cg);
+   cg->getS390OutOfLineCodeSectionList().push_front(outlinedSlowPath);
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, oolFailLabel);
+
+   TR::Node *_callNode = createOutlinedCallNode(cg, node, TR::acall);
+   TR::Register *targetReg2 = NULL;
+   targetReg2 = TR::TreeEvaluator::performCall(_callNode, false, cg);
+   cg->decReferenceCount(_callNode);
+   generateRRInstruction(cg, TR::InstOpCode::LR, node, targetReg, targetReg2);
+
+   // After returning from the helper, branch back to mainline code.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+   // End of OOL
+   outlinedSlowPath->swapInstructionListsWithCompilation();
 
+   // In the mainline, first load the first and second dimensions' lengths into registers.
+   TR::Register *firstDimLenReg = cg->allocateRegister();
+   generateRXInstruction(cg, TR::InstOpCode::LG, node, firstDimLenReg, generateS390MemoryReference(dimsPtrReg, 4, cg));
```

Comment:
I don't know enough about this opcode (multiANewArray), so I guess this is an opportunity for me to learn. What is the `dimsPtrReg` pointing to? It looks like a struct of two 64-bit values? Is this true even on 32-bit/compressedrefs? Could you describe what children of this evaluator represent?

Reply:
Here's an example tree:
```
 n18n     (  0)  treetop                                                                              [     0x3ff7de09e40] bci=[-1,2,13] rc=0 vc=144 vn=- li=2 udi=- nc=1
 n16n     (  2)    multianewarray  jitAMultiANewArray[#36  helper Method] [flags 0x400 0x0 ] (X!=0 sharedMemory )  [     0x3ff7de09da0] bci=[-1,2,13] rc=2 vc=144 vn=- li=2 udi=- nc=3 flg=0x24
 n116n    (  1)      loadaddr  <auto slot -1>[#387  Auto] [flags 0x3 0x0 ]                            [     0x3ff7de86400] bci=[-1,2,13] rc=1 vc=144 vn=- li=- udi=- nc=0
 n17n     (  1)      iconst 2 (X!=0 X>=0 )                                                            [     0x3ff7de09df0] bci=[-1,2,13] rc=1 vc=144 vn=- li=2 udi=- nc=0 flg=0x104
 n15n     (  1)      loadaddr  [[I[#377  Static] [flags 0x18307 0x0 ]                                 [     0x3ff7de09d50] bci=[-1,2,13] rc=1 vc=144 vn=- li=2 udi=- nc=0
```

The first child points to a structure that holds the size of each dimension.
The second child represents the number of dimensions in the multidimensional array.
The third child holds the class of the array.

Reply:
> The first child points to a structure that holds the size of each dimension.

So it is an array of sizes. What is the type of array? i.e. how wide is each filed? Is is 4 bytes? I see we restrict this evaluator to only a two dimensional multianewarray, so just making sure each size field is 4-bytes because we do a 8-byte load here.

Reply:
I can confirm type is an `int` as per the Java lang. spec: https://docs.oracle.com/javase/specs/jls/se8/html/jls-10.html.

The `LG` is incorrect here and will be changed in my next commit.

Reply:
This should no longer be an issue: ff6da2b